File: routes/tasks.py
from flask import Blueprint, flash, redirect, render_template, request, url_for

from auth_helpers import login_required
from extensions import db
from models import Task
from services import current_user_id, paginate_query, parse_iso_date, sanitize_plain, timestamp_now


tasks_bp = Blueprint("tasks", __name__)


@tasks_bp.route("/tasks", methods=["GET", "POST"])
@login_required
def tasks():
    user_id = current_user_id()
    if request.method == "POST":
        task_text = sanitize_plain(request.form.get("task"))
        subject = sanitize_plain(request.form.get("subject"))
        priority = sanitize_plain(request.form.get("priority")) or "medium"
        due_date = parse_iso_date(request.form.get("due"))
        if priority not in {"low", "medium", "high"}:
            priority = "medium"
        if task_text and due_date:
            db.session.add(
                Task(
                    user_id=user_id,
                    task=task_text,
                    subject=subject,
                    priority=priority,
                    due_date=due_date,
                    completed=False,
                    archived=False,
                    created_at=timestamp_now(),
                )
            )
            db.session.commit()
            flash("Task added to your planner.", "success")
        else:
            flash("Task and due date are required.", "error")
        return redirect(url_for("tasks.tasks"))

    status = request.args.get("status", "open")
    priority = request.args.get("priority", "")
    page = request.args.get("page", 1, type=int)
    task_query = Task.query.filter_by(user_id=user_id)
    if status == "completed":
        task_query = task_query.filter_by(completed=True, archived=False)
    elif status == "archived":
        task_query = task_query.filter_by(archived=True)
    else:
        task_query = task_query.filter_by(completed=False, archived=False)
    if priority in {"low", "medium", "high"}:
        task_query = task_query.filter_by(priority=priority)

    paginated_tasks = paginate_query(
        task_query.order_by(Task.due_date.asc(), Task.priority.desc()),
        page,
        12,
    )
    return render_template(
        "tasks.html",
        tasks=paginated_tasks["items"],
        pagination=paginated_tasks,
        status=status,
        priority=priority,
    )


@tasks_bp.route("/tasks/<int:task_id>/complete", methods=["POST"])
@login_required
def complete_task(task_id):
    task = Task.query.filter_by(id=task_id, user_id=current_user_id()).first()
    if not task:
        flash("Task not found.", "error")
        return redirect(url_for("tasks.tasks"))

    task.completed = True
    db.session.commit()
    flash("Task marked complete.", "success")
    return redirect(url_for("tasks.tasks"))


@tasks_bp.route("/tasks/<int:task_id>/<action>", methods=["POST"])
@login_required
def update_task_state(task_id, action):
    task = Task.query.filter_by(id=task_id, user_id=current_user_id()).first()
    if not task:
        flash("Task not found.", "error")
        return redirect(url_for("tasks.tasks"))

    if action == "reopen":
        task.completed = False
        task.archived = False
    elif action == "archive":
        task.archived = True
    elif action == "restore":
        task.archived = False
    elif action == "delete":
        db.session.delete(task)
        db.session.commit()
        flash("Task deleted.", "success")
        return redirect(url_for("tasks.tasks"))
    else:
        flash("Unknown task action.", "error")
        return redirect(url_for("tasks.tasks"))

    db.session.commit()
    flash("Task updated.", "success")
    return redirect(url_for("tasks.tasks"))

File: lib/features/tasks.ts
import { and, asc, desc, eq } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { redirect } from "next/navigation";
import { db } from "@/db";
import { tasks } from "@/db/schema";
import { verifyCsrfToken } from "@/lib/auth/csrf";
import { currentUser } from "@/lib/auth/session";
import { invalidateAcademicAutopilotCache, ownedAcademicSelection } from "@/lib/features/academic";
import { parseIsoDate, sanitizePlain } from "@/lib/text";

const allowedPriorities = new Set(["low", "medium", "high"]);
const allowedActions = new Set(["complete", "reopen", "archive", "restore", "delete"]);

function redirectWith(message: string, type: "error" | "success" = "success"): never {
  redirect(`/tasks?${type}=${encodeURIComponent(message)}`);
}

export async function createTaskAction(formData: FormData) {
  "use server";
  try {
    await verifyCsrfToken(formData);
  } catch {
    redirectWith("Security check failed. Please try again.", "error");
  }
  const user = await currentUser();
  if (!user) redirect("/login");
  const task = sanitizePlain(formData.get("task"));
  const subject = sanitizePlain(formData.get("subject")).slice(0, 128) || null;
  const dueDate = parseIsoDate(formData.get("due"));
  const requestedPriority = sanitizePlain(formData.get("priority"));
  const priority = allowedPriorities.has(requestedPriority) ? requestedPriority : "medium";
  const academic = await ownedAcademicSelection(user.id, formData.get("subject_id"), formData.get("topic_id"));
  if (!task || !dueDate) redirectWith("Task and due date are required.", "error");

  await db.insert(tasks).values({
    userId: user.id,
    subjectId: academic.subjectId,
    topicId: academic.topicId,
    task,
    subject: academic.subjectName ?? subject,
    priority,
    dueDate,
    completed: false,
    archived: false,
    createdAt: new Date(),
  });
  invalidateAcademicAutopilotCache(user.id);
  revalidatePath("/tasks");
  revalidatePath("/autopilot");
  revalidatePath("/");
  redirectWith("Task added to your planner.");
}

export async function updateTaskStateAction(formData: FormData) {
  "use server";
  try {
    await verifyCsrfToken(formData);
  } catch {
    redirectWith("Security check failed. Please try again.", "error");
  }
  const user = await currentUser();
  if (!user) redirect("/login");
  const taskId = Number(formData.get("task_id"));
  const action = sanitizePlain(formData.get("action"));
  if (!allowedActions.has(action)) redirectWith("Unknown task action.", "error");
  const [task] = await db.select().from(tasks).where(and(eq(tasks.id, taskId), eq(tasks.userId, user.id))).limit(1);
  if (!task) redirectWith("Task not found.", "error");

  if (action === "delete") {
    await db.delete(tasks).where(eq(tasks.id, task.id));
    invalidateAcademicAutopilotCache(user.id);
    revalidatePath("/tasks");
    revalidatePath("/autopilot");
    revalidatePath("/");
    redirectWith("Task deleted.");
  }

  const updates: Partial<typeof tasks.$inferInsert> = {};
  if (action === "complete") updates.completed = true;
  if (action === "reopen") {
    updates.completed = false;
    updates.archived = false;
  }
  if (action === "archive") updates.archived = true;
  if (action === "restore") updates.archived = false;

  await db.update(tasks).set(updates).where(eq(tasks.id, task.id));
  invalidateAcademicAutopilotCache(user.id);
  revalidatePath("/tasks");
  revalidatePath("/autopilot");
  revalidatePath("/");
  redirectWith(action === "complete" ? "Task marked complete." : "Task updated.");
}

export function taskStatusPredicate(userId: number, status: string, priority: string) {
  const predicates = [eq(tasks.userId, userId)];
  if (status === "completed") predicates.push(eq(tasks.completed, true), eq(tasks.archived, false));
  else if (status === "archived") predicates.push(eq(tasks.archived, true));
  else predicates.push(eq(tasks.completed, false), eq(tasks.archived, false));
  if (allowedPriorities.has(priority)) predicates.push(eq(tasks.priority, priority));
  return and(...predicates);
}

export function taskOrder() {
  return [asc(tasks.dueDate), desc(tasks.priority)] as const;
}
